ES6+ Syntax & API Separation

walmat opened the pull request.

walmat committed:
es6 rewrite, modularity improvements, and updated readme

walmat committed:
Update index.js

walmat committed:
added babel transformer

walmat committed:
merge upstream

walmat committed:
revert cached git store

walmat committed:
fix entry point and small bugs

walmat committed:
Update README.md

walmat committed:
some bug fixes and examples

walmat committed:
Merge branch 'es6-rewrite' of https://github.com/walmat/stockx-api into es6-rewrite

walmat committed:
added a few examples

walmat committed:
switch to node-fetch

walmat committed:
readded base url

walmat committed:
few fetch bugs fixed

walmat committed:
bug fixes with OAuth

walmat committed:
implemented dotenv to protect sensitive data

walmat committed:
revert to request-promise-native and more bug fixes and qol improvements

walmat committed:
small bug with inclusive range fixed

walmat committed:
implemented singleton and fixed some request-promise issues

walmat committed:
Update README.md

walmat committed:
Update README.md

walmat committed:
updated example

walmat committed:
cleanup duplicate headers

walmat committed:
added better error messages

walmat committed:
added better error messages

matthew1232 committed:
Fixed callback submission

click2install commented on src/apis/stockx/asks/index.js:
```diff
@@ -0,0 +1,140 @@
+    let chainId;
+    let skuUuid;
+    try {
+      ({ chainId, skuUuid } = ask);
```
no need to keep these outside of the `try/catch` scope.
```js
let { chainId, skuUuid } = ask;
```

click2install replied:
I think this was in multiple places, but I left only 1 comment.

walmat committed:
Merge pull request #2 from matthew1232/patch-1

Fixed callback submission

walmat commented:
Manual Checks to test:
- [ ] Can the user successfully login
- [ ] Can the user place an ask
- [ ] Can a user update that previously placed ask
- [ ] Can a user place a bid
- [ ] Can a user update that previously placed bid
- [ ] ..anything else?

click2install commented on src/apis/stockx/asks/index.js:
```diff
@@ -0,0 +1,140 @@
+          authorization: `Bearer ${bearer}`,
+          'content-type': 'application/json',
+        },
+        jar,
```
Given this is duplicated for each request, would be good to abstract it out to a common call, abstracting the request might be a cleaner approach which just takes optional arguments to suppliment the headers, body, etc.

walmat replied:
Good thinking. I think for this PR this is fine, we can do a follow up PR to cleanup and condense.

walmat commented:
@matthew1232 let me know what is left to test, I'll run through a code review of this right now in it's current state.

walmat commented on examples/index.js:
```diff
@@ -0,0 +1,33 @@
+const { username, password } = process.env;
+
+const stockXController = new StockX();
+// if you're using charlesproxy, make sure it's launched and use the controller belo!
```
```suggestion
// if you're using charlesproxy, make sure it's launched and use the controller below!
```
Fix typo

click2install commented on src/apis/stockx/asks/index.js:
```diff
@@ -0,0 +1,140 @@
+      return { chainId, skuUuid };
+    } catch (error) {
+      const err = new Error(`Unable to place ask: ${error.message}`);
+      err.stack = error.stack || {};
```
Given this will be a conumed api, it would be best to not leak stack traces in production environments.

walmat replied:
Yeah I was toying with exposing it, but I think you're right.

click2install replied:
Depending on how you want to proceed, You could add a logger of different levels and permit stack traces in the logs but not back to the consumer for production environments.

walmat commented on package.json:
```diff
@@ -1,9 +1,11 @@
-  "description": "communicate with the unofficial StockX API",
-  "main": "index.js",
+  "version": "1.0.0",
+  "description": "Unofficial StockX API",
```
Come up with a better description. @matthew1232 any ideas?

walmat replied:
Maybe `Unofficial StockX Public API`?

walmat commented on package.json:
```diff
@@ -1,9 +1,11 @@
-  "version": "1.0.3",
-  "description": "communicate with the unofficial StockX API",
-  "main": "index.js",
+  "version": "1.0.0",
```
```suggestion
  "version": "1.0.4",
```
Probably will have to bump this to `1.0.4` since it looks like the most recent version (located on npm) was 1.0.3

walmat replied:
If you want, we can also do a minor version change (1.1.0)

walmat commented on package.json:
```diff
@@ -1,9 +1,11 @@
+  "main": "dist/index.js",
   "scripts": {
+    "build": "babel --no-comments -d dist src/",
+    "start": "node dist/index.js",
```
Probably can remove this script since it won't ever be used..
```suggestion
```

walmat commented on package.json:
```diff
@@ -13,16 +15,43 @@
+    "sneakers",
     "bots",
-    "request"
+    "axios"
```
```suggestion
    "request",
    "promise",
    "es6"
```

matthew1232 replied:
Pretty sure that's a typo, "requuest" should be "request"

walmat replied:
Whoops! Good catch.

walmat commented on src/apis/stockx/asks/index.js:
```diff
@@ -0,0 +1,140 @@
+import { randomInclusive } from '../../../utils';
+
+export default class AsksApi extends Api {
+  // TODO!
```
Will open a follow up issue to implement the remainder of methods. Doesn't necessarily need to be included in this PR.

walmat replied:
Tracked in https://github.com/matthew1232/stockx-api/issues/5

click2install commented on src/apis/stockx/index.js:
```diff
@@ -0,0 +1,205 @@
+      },
+      proxy: this.proxy,
+      bearer: this._bearer,
+      isLoggedIn: this._isLoggedIn,
```
Why maintain these on the two instances of `_context` and `this` and not just `_context`?

walmat replied:
Good catch! This was before the shared data singleton, so I'll make those changes now.

walmat commented on src/apis/stockx/asks/index.js:
```diff
@@ -0,0 +1,140 @@
+      if (!statusCode || (statusCode && statusCode !== 200)) {
+        const err = new Error('Invalid response code!');
+        if (statusCode === 400) {
+          err.message = 'Account on hold!';
```
Is this the best message? Maybe `Account suspended`?

walmat replied:
@matthew1232 lemme know your thoughts

click2install replied:
I am not completely familiar with their api, is `Bad Request` indicative of an existing account that no longer works?

matthew1232 replied:
If the stringified JSON response includes 'account suspended' or whatever the body is, send the error "Account suspended". Same with includes "invalid email" throwing "Invalid email/password", which I had in my previous version.

matthew1232 replied:
The 400 response will happen if the email/password is incorrect as well.

walmat replied:
Ah okay gotcha. I think I’ll amend that back in then, otherwise I’ll just send a generic “Bad Request” message back.

click2install replied:
If that's the case, then when you refactor the "other error handling" in a different PR, you could make a strategy factory that deals with all the 3rd party api status codes and creates `Error`'s for your API.  Keeps it all central, DRY and OCP then.

walmat replied:
Resolved. Added a central error throwing mechanism that includes just enough details to the user while keeping the stack under wrap.

walmat commented on src/apis/stockx/asks/index.js:
```diff
@@ -0,0 +1,140 @@
+    }
+  }
+
+  // TODO!
```
Will open a follow up issue to implement the remainder of methods. Doesn't necessarily need to be included in this PR.

walmat replied:
Done and merged..

walmat commented on src/apis/stockx/bids/index.js:
```diff
@@ -0,0 +1,136 @@
+import { randomInclusive } from '../../../utils';
+
+export default class BidsApi extends Api {
+  // TODO!
```
Will open a follow up issue to implement the remainder of methods. Doesn't necessarily need to be included in this PR.

walmat replied:
Tracked in https://github.com/matthew1232/stockx-api/issues/4

walmat commented on src/apis/stockx/bids/index.js:
```diff
@@ -0,0 +1,136 @@
+      if (!statusCode || (statusCode && statusCode !== 200)) {
+        const err = new Error('Invalid response code!');
+        if (statusCode === 400) {
+          err.message = 'Account on hold!';
```
Same here. Is this the most accurate message we could provide the user?

click2install replied:
Same as above, probably need to understand what triggers a `Bad Request` and if your status message is broad enough to cover all cases.

walmat commented on src/apis/stockx/bids/index.js:
```diff
@@ -0,0 +1,136 @@
+    }
+  }
+
+  // TODO!
```
Will open a follow up issue to implement the remainder of methods. Doesn't necessarily need to be included in this PR.

walmat replied:
Done and merged..

walmat commented on src/apis/stockx/index.js:
```diff
@@ -0,0 +1,205 @@
+   * @param options.currency - The currency used when making requests
+   */
+  constructor(options = {}){
+    const { proxy = null, currency = currencies.USD } = options;
```
Should we default to `USD`?

click2install replied:
If it's documented this would probably be the most common, `options` has the opportunity to override the default, so I see no issue.

matthew1232 replied:
Yes

walmat commented on src/apis/stockx/index.js:
```diff
@@ -0,0 +1,205 @@
+    this.user = new UserApi(this._context);
+  };
+
+    // /**
```
Cleanup the below comments! I think at this point, they can be removed.

click2install commented on src/apis/stockx/user/helpers.js:
```diff
@@ -0,0 +1,131 @@
+      const err = new Error('Invalid response code!');
+      err.status = statusCode || 404;
+      throw err;
+    }
```
This is duplicated quite a bit, refactoring it to a common function would be cleaner.

walmat commented on src/apis/stockx/products/index.js:
```diff
@@ -0,0 +1,81 @@
+    let url = `https://stockx.com/api/browse?&_search=${query}`;
+
+    if (type) {
+      url += `&dataType=${type}`;
```
test searches with data-types

walmat replied:
Works flawlessly, resolving!

walmat commented on src/apis/stockx/products/index.js:
```diff
@@ -0,0 +1,81 @@
+    try {
+      const { currency, headers } = this.data;
+      const { pathname } = new URL(product, 'https://stockx.com');
+      let url = `https://stockx.com/api/products${pathname}?includes=market&currency=${currency}`;
```
Include a boundary check here to make sure pathname exists. They could pass in a URL that isn't for stockX that might throw an error here..

walmat commented on src/apis/stockx/user/helpers.js:
```diff
@@ -0,0 +1,134 @@
+import cheerio from 'cheerio';
```
Look into alternatives for cheerio. If this library will be included in projects, we want to minimize it (bundle size) as much as possible. Cheerio is unnecessarily large if we just need to grab a few form values in one file imho.

walmat replied:
Maybe [hyntax](https://github.com/mykolaharmash/hyntax)? It's a total of 4.7kB compared to cheerio which is 319.3kB (when minified).

matthew1232 replied:
https://www.npmjs.com/package/dom-parser looks like a good option.

walmat replied:
Perfect! Even smaller 🙂

walmat replied:
Implemented [html-dom-parser](https://www.npmjs.com/package/html-dom-parser) which is 3.3kB in size. Resolving...

walmat commented on src/apis/stockx/user/helpers.js:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // Get state and client ID
+    const state = _header.split('state=')[1].split('&')[0];
```
```suggestion
    const [state] = _header.split('state=')[1].split('&');
```

walmat commented on src/apis/stockx/user/helpers.js:
```diff
@@ -0,0 +1,134 @@
+
+    // Get state and client ID
+    const state = _header.split('state=')[1].split('&')[0];
+    const client_id = _header.split('client=')[1].split('&')[0];
```
```suggestion
    const [client_id] = _header.split('client=')[1].split('&');
```

click2install replied:
`URLSearchParams` from `url.serchParams` might be nicer/safer for url parsing/interaction.

walmat commented on src/apis/stockx/user/index.js:
```diff
@@ -0,0 +1,70 @@
+import { getState, submitCredentials, checkStatus } from './helpers';
+
+export default class UserApi extends Api {
+  // todo:
```
Will open a follow up issue to implement the remainder of methods. Doesn't necessarily need to be included in this PR.

walmat replied:
Tracked in https://github.com/matthew1232/stockx-api/issues/6

walmat commented on src/index.js:
```diff
@@ -0,0 +1,3 @@
+import StockX from './apis/stockx';
+
+export { StockX };
```
We could change this to a default export if you want. I don't know your plans for this repository, but if you wanted to do more than just a StockX API, you could export more than just StockX here.

I'm all for moving to a default export right now though, as that's the current direction of the API.

walmat replied:
```suggestion
export default StockX;
```

We will need to go through and update the README to represent a default import, however.

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
```diff
@@ -13,288 +15,363 @@ Or with yarn,
-const stockX = new stockxAPI({
-    proxy: 'proxyhere',
-    currency: 'USD'
+import { StockX } from 'stockx-api';
```
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
```diff
@@ -13,288 +15,363 @@ Or with yarn,
-})
-.then(products => console.log(products))
-.catch(err => console.log(`Error searching: ${err.message}`));
+import { StockX } from 'stockx-api';
```
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
````diff
@@ -13,288 +15,363 @@ Or with yarn,
 ```js
-const stockxAPI = require('stockx-api');
-const stockX = new stockxAPI();
+import { StockX } from 'stockx-api';
````
```suggestion
import StockX from 'stockx-api';
```

walmat commented on README.md:
```diff
@@ -13,288 +15,363 @@ Or with yarn,
+        const product = await stockXController.products.details('https://stockx.com/adidas-yeezy-boost-700-magnet');
 
-    console.log('Placing a bid for ' + product.name);
+        console.log('Placing an bid for ' + product.title);
```
```suggestion
        console.log('Placing a bid for ' + product.title);
```

walmat commented on README.md:
```diff
@@ -13,288 +15,363 @@ Or with yarn,
+
+## License
+
+This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details
```
Write proper LICENSE

walmat replied:
resolved..

walmat commented on README.md:
```diff
@@ -13,288 +15,363 @@ Or with yarn,
+------
+
+## Contributing
+Please read [CONTRIBUTING.md]() for details on our code of conduct, and the process for submitting pull requests to us.
```
Write proper CONTRIBUTING

click2install commented on src/apis/stockx/index.js:
```diff
@@ -0,0 +1,205 @@
+    };
+
+    // sub-module constructors
+    this.products = new ProductsApi(this._context);
```
If the `_context` is given to each class like this, there doesn't seem to be any need for the singleton `SharedContext` class ... unless for encapsulation of the data, in which case it would be best to construct the `SharedContext` here as only the first `Api` with `this._context` is building the singleton, the other `_this.context` calls are ignored as the singleton already exists. If you construct it here and "inject" it into the `Api` classes there is no need for a singleton.

walmat replied:
Resolved.

walmat commented on README.md:
```diff
@@ -13,288 +15,363 @@ Or with yarn,
+
+## Acknowledgements
+
+* **matthew#1232** - *Initial StockX API* - [matthew1232](https://github.com/matthew1232)
```
```suggestion
* **matthew#1232** - *Initial StockX API* - [matthew1232](https://github.com/matthew1232)
* **orion#0001** - *ES6 Rewrite* - [walmat](https://github.com/walmat)
```

walmat reviewed (commented):
First pass done.

click2install reviewed (commented):
Some general observations. Not sure at what stage the PR is at, most comments are w.r.t the code being made more DRY and a few small other observations.

walmat committed:
Apply suggestions from code review

First pass cleanup changes

walmat committed:
fixed shared context, exposed shared context mutations, added shared error logging, and switch from cheerio to html-dom-parser

walmat committed:
fix bug with placing ask/bid

walmat commented:
All manual checks passed.

walmat committed:
updated examples to update ask

walmat committed:
updated description

walmat committed:
fixed broken links in readme, added contributing guidelines

walmat committed:
Update README.md

matthew1232 committed:
Login credential status/general request errors

Added error helper functions for checking login credential submission status + general request errors. Includes an object of the message/status/body so the user can debug easier. These haven't been tested but should work.

matthew1232 committed:
Implementing error handling from utils

matthew1232 committed:
Removed unnecessary error bubbling

matthew1232 committed:
Added deleting bids, implemented checking status from utils

I added the option to delete bids, and implemented the checkStatus from the util errors file. I also removed the error bubbling and added the `body` param to the errors so the errors stay consistent across the board.

matthew1232 committed:
Updated error handling, removed unnecessary error bubbling, adding removing asks

walmat committed:
Merge pull request #4 from matthew1232/patch-3

Login credential status/general request errors

walmat committed:
Merge pull request #5 from matthew1232/patch-4

Implementing error handling from utils and changed checkStatus to submitCallback

walmat committed:
Merge pull request #7 from matthew1232/patch-6

Removed unnecessary error bubbling

matthew1232 committed:
Include 304 as accepted status code

matthew1232 committed:
Update src/apis/stockx/asks/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/asks/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/asks/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/asks/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

walmat committed:
Merge pull request #10 from matthew1232/patch-10

Include 304 as accepted status code

matthew1232 committed:
Import error handling into utils

walmat committed:
Merge pull request #11 from matthew1232/patch-11

Import error handling into utils

matthew1232 committed:
Import utils from index

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update index.js

walmat committed:
Merge pull request #9 from matthew1232/patch-9

Updated error handling, removed unnecessary error bubbling, adding re…

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

walmat reviewed (commented):
Opened remaining separate issues and resolved. Should re-validate manual checks before merging!

matthew1232 committed:
Update src/apis/stockx/bids/index.js

Co-Authored-By: Matthew Wall <matthew.wallt@gmail.com>

walmat committed:
Merge pull request #8 from matthew1232/patch-8

Added deleting bids, implemented checking status from utils

matthew1232 committed:
Add error handling from utils

Allowed objects from products.search, fixed error handling and removed error bubbling that was unneeded.

walmat committed:
Merge pull request #13 from matthew1232/patch-8

Add error handling from utils

matthew1232 committed:
Remove renamed variables (fatal)

walmat committed:
Merge pull request #18 from matthew1232/patch-13

Remove renamed variables (fatal)
